### PR TITLE
[fsdp, fully_async] feat: add Qwen3-VL-30B-A3B fully async GRPO training script on geo3k

### PR DESCRIPTION
**Pull request by zhihaofang1017**
### What does this PR do?
Fix 2 bugs related to launch **Qwen3-VL-30B-A3B**, one bug caused accuracy issues, and the other caused training errors.
**Bug1：**
We find that the `reward` is not trending upwards, and the `ppo-kl `is several orders of magnitude lower than the GPU. We need to remove the line `if not self.training` from the `verl/models/transformers/npu_patch.py`. After the change, both the accuracy and reward are now trending upwards, and the previously mentioned `ppo-kl `value is now reasonable. (Green represents the modified version, blue represents the unmodified open-source version)
screenshot-20260427-103501

**Bug2：**
Modifying the device settings in `fast_pos_embed_interpolate` via `verl/models/transformers/monkey_patch.py.`
This is a known issue related to transformers. Qwen3.5 also avoids this problem in this way. 

After fix these problems, we add a fully async GRPO training script for **Qwen3-VL-30B-A3B** on the geo3k dataset under `verl/experimental/fully_async_policy/shell/`.

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: ...
- [ ] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `vllm_omni`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `fully_async`, `one_step_off`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test
**Environment**
Tested on Ascend NPU. Refer to [Ascend Quickstart](https://github.com/volcengine/verl/blob/main/docs/ascend_tutorial/quick_start/ascend_quick_start.rst) for full installation instructions. Core versions:
Software | Version
-- | --
CANN | 8.5.0
torch | 2.8.0
torch_npu | 2.8.0
vllm | 0.13.0
vllm-ascend | 0.13.0
transformers | 4.57.6

**API and Usage Example**
No API changes. To run:
```shell
bash verl/experimental/fully_async_policy/geo3k_qwen3vl_30b_a3b_fsdp2_32_32.sh`
```

**Results**
reward: I have already shown it before.
acc:
screenshot-20260427-103704

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/verl-project/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [ ] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

**Comment by ArronHZG** [comment]
> Modifying the device settings in fast_pos_embed_interpolate via verl/models/transformers/monkey_patch.py.
This is a known issue related to transformers. Qwen3.5 also avoids this problem in this way.

Please specify the source of the modification in code: fast_pos_embed_interpolate
If you have any `transformer` related issues, and `transformer` version,  you can raise them here.

File: verl/models/transformers/qwen3_vl.py [hunk 1]
```diff
@@ -29,6 +29,68 @@
...
+        weight_list, dtype=self.pos_embed.weight.dtype, device=device
+    )
+    pos_embeds = self.pos_embed(idx_tensor) * weight_tensor[:, :, None]
+    patch_pos_embeds = pos_embeds[0] + pos_embeds[1] + pos_embeds[2] + pos_embeds[3]
+
+    patch_pos_embeds = patch_pos_embeds.split([h * w for h, w in zip(grid_hs, grid_ws)])
+
+    patch_pos_embeds_permute = []
+    merge_size = self.config.spatial_merge_size
+    for pos_embed, t, h, w in zip(patch_pos_embeds, grid_ts, grid_hs, grid_ws):
```
Comment (wucong25) [inline verl/models/transformers/qwen3_vl.py]: What specific issue does this patch fix?

File: verl/models/transformers/npu_patch.py [hunk 2]
```diff
@@ -281,8 +281,7 @@ def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
         routing_weights = routing_weights / routing_weights.sum(dim=-1, keepdim=True)
         routing_weights = routing_weights.to(router_logits.dtype)
         hidden_states = hidden_states.reshape(batch_size, -1, self.hidden_size)
-        if not self.training:
```
Comment (wucong25) [inline verl/models/transformers/npu_patch.py:284]: Have you verified whether the previous scenarios required going into this branch?

Comment (zhihaofang1017) [reply verl/models/transformers/npu_patch.py:284]: Yes, I have verified that Qwen3-vl-moe will go to this npu-patch branch

**Review by wuxibin89 — DISMISSED** [review]

**Comment by wuxibin89** [comment]
Please format code before commit: https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md#code-linting-and-formatting

**Comment by zhihaofang1017** [comment]
> Please format code before commit: https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md#code-linting-and-formatting

I have formatted the code

**Review by wucong25 — APPROVED** [review]
